// slonik/dist/tokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnnestToken = exports.TimestampToken = exports.QueryToken = exports.ListToken = exports.JsonToken = exports.JsonBinaryToken = exports.IntervalToken = exports.IdentifierToken = exports.FragmentToken = exports.DateToken = exports.ComparisonPredicateToken = exports.BinaryToken = exports.ArrayToken = void 0;
exports.ArrayToken = 'SLONIK_TOKEN_ARRAY';
exports.BinaryToken = 'SLONIK_TOKEN_BINARY';
exports.ComparisonPredicateToken = 'SLONIK_TOKEN_COMPARISON_PREDICATE';
exports.DateToken = 'SLONIK_TOKEN_DATE';
exports.FragmentToken = 'SLONIK_TOKEN_FRAGMENT';
exports.IdentifierToken = 'SLONIK_TOKEN_IDENTIFIER';
exports.IntervalToken = 'SLONIK_TOKEN_INTERVAL';
exports.JsonBinaryToken = 'SLONIK_TOKEN_JSON_BINARY';
exports.JsonToken = 'SLONIK_TOKEN_JSON';
exports.ListToken = 'SLONIK_TOKEN_LIST';
exports.QueryToken = 'SLONIK_TOKEN_QUERY';
exports.TimestampToken = 'SLONIK_TOKEN_TIMESTAMP';
exports.UnnestToken = 'SLONIK_TOKEN_UNNEST';
